cluster-bootstrap/src/test/scala/akka/management/cluster/bootstrap/contactpoint/ClusterBootstrapDiscoveryBackoffIntegrationSpec.scala

```diff
@@ -0,0 +1,161 @@
+
+    val baseTime = System.currentTimeMillis()
+    case class DiscoveryRequest(time: Long, attempt: Int, res: Future[_]) {
+      override def toString = s"DiscoveryRequest(${(time - baseTime).millis}, $attempt, $res)"
+    }
+    val resolveProbe = TestProbe()(systemA)
+
+    // prepare the "mock DNS" - this resolves to unreachable addresses for the first three
+    // times it is called, thus testing that discovery is called multiple times and
+    // that formation will eventually succeed once discovery returns reachable addresses
+
+    @volatile var called = 0
+    val name = "system.svc.cluster.local"
+
+    MockDiscovery.set(name, { () =>
+      called += 1
+
+      val res =
+        if (called < 4)
+          Future.failed(new Exception("Boom! Discovery failed, was rate limited for example..."))
+        else
+          Future.successful(Resolved(name,
+              List(
+                ResolvedTarget(clusterA.selfAddress.host.get, contactPointPorts.get("A")),
+                ResolvedTarget(clusterB.selfAddress.host.get, contactPointPorts.get("B"))
+              )))
+
+      resolveProbe.ref ! DiscoveryRequest(System.currentTimeMillis(), called, res)
+      res
+    })
+
+    "start listening with the http contact-points on 3 systems" in {
+      def start(system: ActorSystem, contactPointPort: Int) = {
+        import system.dispatcher
+        implicit val sys = system
+        implicit val mat = ActorMaterializer()(system)
+
+        val bootstrap = ClusterBootstrap(system)
+        val routes = new HttpClusterBootstrapRoutes(bootstrap.settings).routes
+        bootstrap.setSelfContactPoint(s"http://127.0.0.1:$contactPointPort")
+        Http().bindAndHandle(RouteResult.route2HandlerFlow(routes), "127.0.0.1", contactPointPort)
+      }
+
+      start(systemA, contactPointPorts("A"))
+      start(systemB, contactPointPorts("B"))
+    }
+
+    "poll discovery in exponentially increasing backoffs until eventually joining the returned nodes" in {
+      bootstrapA.discovery.getClass should ===(classOf[MockDiscovery])
+
+      bootstrapA.start()
+      // bootstrapB.start() // deliberately not discovering from this node, however its management port is available
+
+      val pA = TestProbe()(systemA)
+      clusterA.subscribe(pA.ref, classOf[MemberUp])
+
+      pA.expectMsgType[CurrentClusterState]
+      val up1 = pA.expectMsgType[MemberUp](45.seconds)
+      info("" + up1)
+
+      val timeZero = System.currentTimeMillis()
+      val times = (1 to 6) map { _ =>
+        val lookup = resolveProbe.expectMsgType[DiscoveryRequest]
+        lookup.time
+      }
+      // ends up being very flaky:
+      //      (times(1) - times(0)).toInt shouldBe 2000 +- 500
+      //      (times(2) - times(1)).toInt shouldBe 4000 +- 500
+      //      // so it backs off, later it shall reset to "normal" intervals:
+      //      (times(5) - times(4)).toInt shouldBe 1000 +- 500
+
+      called >= 6 shouldBe true
```

Comment:
I think one can write something like this: `called should >= (6)` for better error messages

Reply:
it's good to only test the number of calls at the end of the test and not try to check intervals, but I think it can be somewhat more specific so that the test would fail without the new backoff implementation.  I think that currently it will pass with old implementation also? i.e. some check that called is also <= X

Reply:
It's a bit fuzzy to put a <  on it since the latter requests could happen a few more or less... by the delays we increased the number of calls since they are failures.. not sure how to better assert -- and the timing assertions are painful on travis :\

Reply:
Would this be a good opportunity to try out the ExplicitlyTriggeredScheduler (https://github.com/akka/akka-http/pull/1221/files)?